[MM-40825] Account setting to always land users at the newest messages in a channel

kyeongsoosoo opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
1. Added an advanced_setting option `unreadScrollPosition`.
2. If an option is set to `start_from_newest`, recent posts are loaded together when unread posts are loaded.
3.  If an option is set to `start_from_newest`, scroll is set to newest message.
4. If an option is set to `start_from_newest` and scroll is at bottom, `unreadWithBottomStart` toast is shown.
5. `unreadWithBottomStart` toast action moves scroll to the oldest unread message.

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
--> 
Fixes https://github.com/mattermost/mattermost-server/issues/19278
Jira ticket: https://mattermost.atlassian.net/browse/MM-40825

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->
- Has server changes (please link here)
- Has mobile changes (please link here)

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->


https://user-images.githubusercontent.com/59413880/170651884-dff3bf33-0d98-480c-aef9-bad7ed942d81.mov



#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Added an advanced_setting option `unreadScrollPosition`
```

mattermod commented:
Hello @kyeongsoosoo,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

jasonblais commented:
Thanks @kyeongsoosoo for your contribution! :tada:

esethna commented:
Awesome, great work @kyeongsoosoo! Will defer to dev for code review

hmhealey commented on components/post_view/post_list_virtualized/post_list_virtualized.jsx:
```diff
@@ -490,6 +496,13 @@ export default class PostList extends React.PureComponent {
             };
         }
 
+        if (this.props.atLatestPost && this.props.shouldStartFromBottomWhenUnread) {
```
Why do we need to check `atLatestPost` here? Wouldn't checking `shouldStartFromBottomWhenUnread` be enough?

kyeongsoosoo replied:
Thanks. `shouldStartFromBottomWhenUnread` would be enough.

hmhealey commented on packages/mattermost-redux/src/actions/posts.ts:
```diff
@@ -776,21 +777,34 @@ export function getPosts(channelId: string, page = 0, perPage = Posts.POST_CHUNK
             posts = await Client4.getPostsUnread(channelId, userId, DEFAULT_LIMIT_BEFORE, DEFAULT_LIMIT_AFTER, fetchThreads, collapsedThreadsEnabled, collapsedThreadsExtended);
+
+            if (posts.next_post_id && shouldLoadRecent) {
+                recentPosts = await Client4.getPosts(channelId, 0, Posts.POST_CHUNK_SIZE / 2, fetchThreads, collapsedThreadsEnabled, collapsedThreadsExtended);
```
What's the motivation for loading two separate chunks of posts here instead of just loading the most recent posts? Is it so that we can still go back to the unread posts?

kyeongsoosoo replied:
>  Is it so that we can still go back to the unread posts?

Yes. I loaded two separate chunks to handle the logic related to "go back to the unread posts".

hmhealey commented on components/toast_wrapper/toast_wrapper.jsx:
```diff
@@ -120,31 +130,45 @@ class ToastWrapper extends React.PureComponent {
+            (props.lastViewedBottom >= props.latestPostTimeStamp) &&
+            prevState.atBottom === null && props.atBottom
+        ) {
+            showUnreadWithBottomStartToast = true;
```
Just to make sure I understand this, we want to show the "scroll up to unreads" toast if we loaded the channel at the bottom and the channel is unread, is that correct? I'm not entirely sure what the 3rd and fourth lines of the if condition are doing

kyeongsoosoo replied:
I was intended to show the "scroll up to unreads" toast if scroll is at the bottom, but I found checking these two lines are unnecessary. I removed these lines.

+) I added a line that checks `props.shouldHideNewMessageIndicator` to hide the toast when scroll reached  the unread posts.

hmhealey reviewed (commented):
Sorry for the delay in reviewing this. It's a big change to a complicated and sensitive part of the code base, so I wanted to make sure I had time to understand it.

I still have to give this a test, but this logic looks pretty sound to me. I have a few questions, but they're just to make sure I understand 100% of what's happening here

hmhealey commented on components/post_view/post_list/index.js:
```diff
@@ -43,14 +43,20 @@ function makeMapStateToProps() {
+        //Until scroll reaches to unread section, there are more than two orders.
+        //If scroll reaches to unread section, orders are merged and length becomes 1.
+        const orders = state.entities.posts.postsInChannel[channelId] ?? [];
+        const shouldHideNewMessageIndicator = shouldStartFromBottomWhenUnread && orders.length > 1;
```
Yep, that's what i was hoping for

hmhealey reviewed (approved):
Thanks for the change and for answering my questions. Just one small thing I'd still like changed, but otherwise, this looks excellent

hmhealey commented:
I'm noticing a couple things with the behaviour of this feature that could be a bit odd, but I'm not sure if they're actually issues. 

1. The "scroll to unreads" toast isn't displayed if you mark a channel as unread a few pages back, switch away from that channel, and then switch back. That's happening because the New Messages line is in the loaded chunk, even if it's a few hundred posts back. I liked the idea of controlling the toast based on whether or not the New Messages line was present because of how simple that is, but we might have to check where it is which would be trickier.
2. Scrolling manually up to the New Messages line doesn't dismiss that toast either. It only disappears if you click on it or change channels.

@esethna Any thoughts on these? I'm not sure if those would be considered bugs here or if they're acceptable behaviour

esethna commented:
@hmhealey sorry I'm not sure I'm understanding #1

Re # 2, good point. I think we should dismiss the toast once the user scrolls up to the new message line when this setting is enabled. Similar to how the toast is dismissed when the user reaches the bottom of the channel today (ie with this new setting disabled)

cc// @kyeongsoosoo

kyeongsoosoo commented:
issue # 1 sounds like a bug and I agree that it would be better to fix issue # 2.
I will look into it and fix these isssues!

Thanks for the comment @esethna , @hmhealey .

kyeongsoosoo commented:
@hmhealey 
I added `isNewMessageLineReached` state to `post_list_virtualized`.
With this state, we can check whether the new message line is displayed in the view.

`unreadWithBottomStartToast` is now shown based on `isNewMessageLineReached` instead of `shouldHideNewMessageIndicator` prop.
I checked that this change fixed the issue and works well with marking a channel as unread.

`unreadWithBottomStartToast` is dismissed when the new message line is displayed in the view.

https://user-images.githubusercontent.com/59413880/173783970-4ca6c160-9ee1-45bd-9ce8-fe28ab9fc761.mov

esethna commented:
@hmhealey please take another look here!

cc// @jgilliam17 to start QA testing

hmhealey commented on components/post_view/post_list_virtualized/post_list_virtualized.jsx:
```diff
@@ -305,6 +307,12 @@ export default class PostList extends React.PureComponent {
         });
     };
 
+    toggleIsNewMessageLineReached = () => {
```
Somewhat nitpicking, but maybe we could call this `onNewMessageLineReached`? The name toggle makes it sound a bit like we're going to be turning this state on and off when you're reading the code for `onItemsRendered`

hmhealey reviewed (commented):
Great idea. I hadn't considered that we could just have the virtualized list tell us when the New Messages line becomes visible. That seems much more straightforward than what I expected we'd need to do.

From playing around with this briefly, this fixes the weird cases I had found. I'm thinking we should probably add E2E cases for this new behaviour as well though. There's already some in `e2e/cypress/tests/integration/toast` that test specifically the behaviour of the toast, so it would be good to have some for this behaviour as well. They'd be better at testing real-world behaviour than the component tests on their own.

hmhealey commented:
Also, sorry for dumping those conflicts on you. I didn't even think how the migration of that component to TypeScript (in https://github.com/mattermost/mattermost-webapp/pull/10426) would affect this PR. Let me know if you need any help resolving those since there's a chance that those are complicated depending on which lines are conflicting

hmhealey commented:
/e2e-tests

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/208983

kyeongsoosoo commented:
@hmhealey 
I  resolved the conflicts and added some e2e tests!

kyeongsoosoo commented:
> Also, sorry for dumping those conflicts on you. I didn't even think how the migration of that component to TypeScript (in #10426) would affect this PR. Let me know if you need any help resolving those since there's a chance that those are complicated depending on which lines are conflicting

I checked that this feature works fine after resolving the conflicts.
Please let me know if I missed something that should be checked.

esethna commented:
Thanks @kyeongsoosoo!

@hmhealey @koox00 please re-review when you have a chance!

hmhealey reviewed (approved):
Thanks for adding those tests!

hmhealey commented:
/e2e-tests

I guess I also should've waited for those conflicts to be resolved before running these. Whoops

koox00 reviewed (approved):
Thank you @kyeongsoosoo much appreciated 💯 !!!

jgilliam17 commented:
/e2e-test

jgilliam17 reviewed (changes requested):
Thanks @kyeongsoosoo 
I suggested few changes to test case IDs below.
Seeing few issues.
1. Unable to dismiss Jump to unreads toast by clicking on the X or using `esc`
![Screenshot 2022-06-23 121718](https://user-images.githubusercontent.com/52937121/175355813-69e2ff94-1bf8-4a26-ad0d-a7d50f19ddc4.png)

2. The new messages count is sometimes wrong (Chrome Win10). In the screenshot I added 30 posts exactly, but toast showed 5. New message line was not visible since it was correctly placed above 1st of 30 new posts.

![Screenshot 2022-06-23 115545](https://user-images.githubusercontent.com/52937121/175356082-62b7f357-6b31-48bd-b9e1-e06106a415f7.png)

3. When marking the oldest message as unread, the message count that is shown at mark as unread time and one that user sees when coming back to the channel differs by few posts
![Screenshot 2022-06-23 130403](https://user-images.githubusercontent.com/52937121/175356778-03bdf9fd-e728-44af-9043-3444260abaa7.png)
![Screenshot 2022-06-23 130444](https://user-images.githubusercontent.com/52937121/175356790-80e50c62-129f-4304-8913-41c78644f1be.png)

4. I've also seen the toast on new channel with single post and showing a wrong date. We shouldn't show any toast in this case
Steps: 
Create a new channel
Mention nonmember user in a post
Add them to channel  
View this channel as new user
Observed: unexpected toast with wrong date
![Screenshot 2022-06-23 115124](https://user-images.githubusercontent.com/52937121/175357022-b2e0841b-6bcc-42e2-b162-c6cb7ca51914.png)

jgilliam17 commented:
I'm not sure why my comment posted twice, sorry about the spam

kyeongsoosoo commented:
Thanks for the QA, @jgilliam17. 
I resolved the issue # 1 and # 4. 
Testing # 2 and # 3, I found a weird point.

I used the Cypress and send 260 messages to the test channel using `postMessageAs` command.
Regardless of which `Scroll position when viewing an unread channel` option is selected, unread toast displayed wrong number when test channel is selected.
I also tested with current master branch and result was same.
<img width="863" alt="스크린샷 2022-06-26 오전 1 00 08" src="https://user-images.githubusercontent.com/59413880/175781453-9028380b-4916-4491-8254-381161e71a41.png">

I'm not sure this is a bug or I did something wrong.
Would you check this, please?


+) 
When I delayed the `postMessageAs` command, there was no problem.
```js
for (let index = 0; index < 260; index++) {
    cy.postMessageAs({sender: otherUser, message: `This is an old message [${index}]`, channelId: testChannelId});
    cy.wait(TIMEOUTS.ONE_HUNDRED_MILLIS);
}
```
<img width="868" alt="스크린샷 2022-06-26 오전 1 17 41" src="https://user-images.githubusercontent.com/59413880/175782109-8b43b209-950f-4e92-a598-e2a80d7e430e.png">

esethna commented:
@jgilliam17 can you help check on the above, thanks! ^

jgilliam17 commented:
Thanks @kyeongsoosoo I'll take a look at this issue today

jgilliam17 commented:
@kyeongsoosoo I am not able to reproduce the discrepancy in the number on the new messages on the unread toast on the master, it seems to be only present on the PR test server. 
I've asked @furqanmlk to also take a look, appreciate your patience  while we investigate. Thanks again 🙂

kyeongsoosoo commented:
Thanks! 😊 @jgilliam17

kyeongsoosoo commented:
I accidentally closed the PR, so I reopened it. Sorry for the mistake.

kyeongsoosoo commented:
@jgilliam17 I've fixed the suggested issues!
The issue I mentioned is not reproduced after I updated the branch and fixed the unread message counting logic. 
Maybe further investigation would not be needed.

+) 
I added an up-arrow icon.

<img width="1489" alt="스크린샷 2022-06-30 오후 3 59 07" src="https://user-images.githubusercontent.com/59413880/176614941-aef96422-5574-4f37-98c2-4003990bed5c.png">

jgilliam17 commented:
/e2e-test

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/213917

jgilliam17 reviewed (changes requested):
@kyeongsoosoo 
Manually re-tested, looks good.
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/213917-8e2ba12-onprem-ent-webapp-pr-10402-mattermost/mm-ee-test8e2ba12/mochawesome.html) shows few toast related failures. Can you please take a look and fix? Thanks 🙂

kyeongsoosoo commented:
@jgilliam17 
Fixed e2e failure cases!

jgilliam17 commented:
/e2e-test

jgilliam17 reviewed (approved):
Thanks @kyeongsoosoo 
[E2E report ](https://mattermost-cypress-report.s3.amazonaws.com/214095-c7f34a6-onprem-ent-webapp-pr-10402-mm-ee-test:c7f34a6/mochawesome.html) looks good, no PR related failures.
<img width="1045" alt="Screen Shot 2022-07-01 at 3 46 03 PM" src="https://user-images.githubusercontent.com/52937121/176960440-daf74c9d-b164-4716-87f5-5d9de68e45fb.png">
<img width="1045" alt="Screen Shot 2022-07-01 at 3 45 50 PM" src="https://user-images.githubusercontent.com/52937121/176960550-9d597e8f-90c7-4f2c-a1a3-8ac4a9b2f6d4.png">

jgilliam17 commented:
@koox00 Can you please check and approve the workflow and merge? Thanks 🙂

mattermod commented:
Will try to auto merge this PR once all tests and checks are passing. This might take up to an hour.

mattermod commented:
Trying to auto merge this PR.

mattermod commented:
Pull Request successfully merged
SHA: db9c10bb1fcc4b9be8357160ee79d127e6d5e83d
